lib/membrane_mp4/demuxer/isom.ex

```diff
@@ -11,4 +20,255 @@ defmodule Membrane.MP4.Demuxer.ISOM do
+      put_in(state, [:buffered_samples, track_id], samples)
+    end)
+  end
+
+  defp get_buffer_actions(samples) do
+    Enum.map(samples, fn {buffer, track_id} ->
+      {:buffer, {Pad.ref(:output, track_id), buffer}}
+    end)
+  end
+
+  defp parse_header(data) do
+    case Container.Header.parse(data) do
+      {:ok, header, _rest} -> header
+      {:error, :not_enough_data} -> nil
+    end
+  end
+
+  defp can_read_data_box?(state) do
+    Enum.all?(@header_boxes, &Keyword.has_key?(state.boxes, &1)) and
+      ((state.last_box_header != nil and state.last_box_header.name == :mdat) or
+         Keyword.has_key?(state.boxes, :mdat))
+  end
+
+  defp get_track_notifications(state) do
+    state.sample_data.sample_tables
+    |> Enum.map(fn {track_id, table} ->
+      content = table.sample_description.content
+      {:notify, {:new_track, track_id, content}}
+    end)
+  end
+
+  defp get_caps(state) do
+    state.sample_data.sample_tables
+    |> Enum.map(fn {track_id, table} ->
+      caps = %Membrane.MP4.Payload{
+        content: table.sample_description.content,
+        timescale: table.timescale,
+        height: table.sample_description.height,
+        width: table.sample_description.width
+      }
+
+      {:caps, {Pad.ref(:output, track_id), caps}}
+    end)
+  end
+
+  @impl true
+  def handle_pad_added(:input, _ctx, state) do
+    {:ok, state}
+  end
+
+  def handle_pad_added(_pad, _ctx, %{all_pads_connected?: true}) do
+    raise "All tracks have corresponding pad already connected"
+  end
+
+  def handle_pad_added(Pad.ref(:output, track_id), ctx, state) do
+    all_pads_connected? = all_pads_connected?(ctx, state)
+
+    {actions, state} =
+      if all_pads_connected? do
+        {buffer_actions, state} = flush_samples(state, track_id)
+        maybe_caps = if state.sample_data != nil, do: get_caps(state), else: []
+        maybe_eos = if state.end_of_stream?, do: get_end_of_stream_actions(ctx), else: []
+
+        {maybe_caps ++ buffer_actions ++ maybe_eos, state}
+      else
+        {[], state}
+      end
+
+    {{:ok, actions}, %{state | all_pads_connected?: all_pads_connected?}}
+  end
+
+  @impl true
+  def handle_end_of_stream(:input, _ctx, %{all_pads_connected?: false} = state) do
+    {:ok, %{state | end_of_stream?: true}}
+  end
+
+  def handle_end_of_stream(:input, ctx, %{all_pads_connected?: true} = state) do
+    {{:ok, get_end_of_stream_actions(ctx)}, state}
+  end
+
+  defp all_pads_connected?(_ctx, %{sample_data: nil}), do: false
+
+  defp all_pads_connected?(ctx, state) do
+    tracks = 1..state.sample_data.tracks_number |> Enum.to_list()
+
+    pads =
+      Enum.filter(ctx.pads, &match?({Pad.ref(:output, _id), _data}, &1))
+      |> Enum.map(fn {Pad.ref(:output, pad_id), _data} -> pad_id end)
+      |> Enum.sort()
+
+    Enum.each(pads, fn pad ->
+      if pad not in tracks do
+        raise "An output pad connected with #{pad} id, however no matching track exists"
+      end
+    end)
```

Comment:
We execute this function each time a pad is connected, so that's actually true. However, if a wrong pad is connected, but we haven't read the `moov` box yet (containing track definitions), we don't really now if any of the pads are incorrect.

Reply:
So we can do that only when moov is there and when moov arrives

Reply:
exactly